### ISSUE TITLE
np.loadtxt fails on empty files (migrated from Trac #1752)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3302]
Original ticket http://projects.scipy.org/numpy/ticket/1752
Reported 2011-03-02 by atmention:WeatherGod, assigned to unknown.

loadtxt() should return an empty array instead of raising an exception for an empty (but existing!) file.  This behavior would be consistent with Matlab.

```
import numpy as np
from StringIO import StringIO

a = np.loadtxt(StringIO(""))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/bvr/Programs/numpy/numpy/lib/npyio.py", line 748, in loadtxt
    raise IOError('End-of-file reached before encountering data.')
IOError: End-of-file reached before encountering data.
```

This has occurred since version 1.2, I believe.  The current behavior is an improvement over previous versions, where numpy would hang on empty files.  Therefore, I am filing this as an enhancement request.

**Comment by thouis** [comment]
Comment in Trac by trac user dynetrekk, 2011-03-26

I agree that the feature request would be nice. I also attach a possible patch, which will simply make loadtxt return an empty array.
    diff -r 2763b87dd7e8 numpy/lib/npyio.py
    --- a/numpy/lib/npyio.py    Fri Mar 25 22:37:19 2011 -0600
    +++ b/numpy/lib/npyio.py    Sat Mar 26 13:10:36 2011 +0100
    atmention:atmention: -745,7 +745,10 atmention:atmention:
             while not first_vals:
                 first_line = fh.readline()
                 if not first_line: # EOF reached
    -                raise IOError('End-of-file reached before encountering data.')
    +                # Break out of the loop here, so that we return an empty array.
    +                first_line = ''
    +                first_vals = []
    +                break
                 first_vals = split_line(first_line)
             N = len(usecols or first_vals)

**Comment by thouis** [comment]
Comment in Trac by atmention:dhomeier, 2011-03-26

Tested in combination with #1562 and appears to work. Funnily, it seems a scalar can have 0 dimensions, but an empty array must have at least 1?

**Comment by thouis** [comment]
Attachment in Trac by atmention:dhomeier, 2011-03-26: [1752test.patch](http://projects.scipy.org/numpy/attachment/ticket/1752/1752test.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:dhomeier, 2011-03-26

It would need to be changed again to work with the proposed patch for #1616, and also test_io.py needs to be changed.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-04-02

This looks like the correct behavior. #1616 needs a change to not break this, this can go in.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-04-03

Fixed in commit:e340e66. Thanks all.
